### PR TITLE
Swift: add location and visibility support to TSP diagnostics

### PR DESCRIPTION
**Pull request by redsun82**
This required abandoning the original integration of JSON diagnostics into binlog.

The problem is that it is quite hard to add any kind of metadata to binlog entries without changing its code, and deserializing with `mserialize` is a bit involved. A thing I tried was to use a `\0` separated format to hide additional data from the binlog message, but the placeholder count check does not go beyond `\0`. Another problem is that when wanting to avoid double evaluation of logging macro arguments one cannot really add a separate "diagnose" step easily to the surface macros.

The proposed solution consists in adding diagnostics emission as a further step at macro invocation.

This allows to forgo the weird category to diagnostic mapping, and now a diagnostics emission simply happens when a special implementation detail diagnostic wrapper type is given as the first argument after the log format string. 

Now locations can be introduced locations by writing for example
```cpp
DIAGNOSE_ERROR(internalError.withLocation("foo.swift", 32), "ouch!");
```

As a nice by-product, `fmt::format` and `fmt::print` are now available without having to wait for C++20 and C++23 respectively.

Moreover, visibility customization was added via a `SwiftDiagnostic::Visibility` enum flag with bitwise operations.

### CODE REVIEW

**Comment by redsun82** [comment]
Thinking a bit more about it, I prefer going back to explicit `DIAGNOSE_ERROR` macros, rather than the implicit magic diagnostic emission on special log arguments, keeping that as a hidden implementation detail behind a special wrapper type. [Explicit is better than implicit](https://peps.python.org/pep-0020/) :slightly_smiling_face:

File: swift/logging/SwiftLogging.h [hunk 1]
```diff
@@ -228,4 +224,54 @@ class Logger {
...
+
+// we intercept this binlog plumbing function providing better overload resolution matches in
+// case the first non-format argument is a diagnostic source, and emit it in that case with the
+// same timestamp
+namespace binlog::detail {
+template <typename Writer, size_t N, typename... T>
+void addEventIgnoreFirst(Writer& writer,
+                         std::uint64_t eventSourceId,
+                         std::uint64_t clock,
+                         const char (&format)[N],
```
Comment (sashabu) [inline swift/logging/SwiftLogging.h:241]: Nit: Can we make this a `const char*` to reduce the number of instantiations? (We don't use the bound anywhere that I can see.)

**Review by sashabu — COMMENTED** [review]
Overall, my feeling is that this is too complex. We have to remember that this is a small project and that the complexity of implementation (that we also have to maintain) won't necessarily pay off in callers' savings.

First, I'll note down the alternatives we discussed on Zoom:
 1. Have `DIAGNOSE_WITH_LEVEL` make two separate calls (`LOG_WITH_LEVEL` and `diagnose`). We rejected this option on the call due to it evaluating arguments (which potentially have side effects) twice. On further reflection, I actually think this is fine (as long as we document it): it's a logging macro, and having side effects in log arguments is unusual (so risk is small) and harms call-site readability (so we don't lose any useful functionality).
 2. Fork Binlog and add support for explicitly passing an `std::source_location` so we can make the wrapper a (template) function instead of a macro (allowing us to evaluate arguments once, at the call-site). (And try to upstream the changes. :) ) We rejected this because of:
  * Maintainability. On further reflection, I think the maintainability is actually *easier* with a fork that has diff/merge support: overloading an internal function as in this PR could break *silently* if we ever upgrade Binlog.
  * Complexity, particularly around constructing a per-call-site EventSource (which probably necessitates making the source location a template parameter (and since `std::source_location` has private members, we'd have to roll our own).

There's also a variation on (1):
  3. Make *callers* do the two calls. This solves the reevaluation issue by pushing it to callers who care. It also gives callers more flexibility. Maybe we'll want to put different information in diagnostics and logs at some point? E.g. a stack trace in the log but not the diagnostic, or even more technical wording in the log.
  * Is this a problem for internal error diagnostics? (Thinking both in terms of making it easy to *remember* adding a diagnostic and not littering the code base with duplicate log lines everywhere.)
  * Are we worried about the copy-pasting bringing accidental inconsistencies?

I'm leaning towards option (1) or (3) if we think they cover our use cases well enough.

That said, I'm open to reconsidering this PR.

File: swift/logging/SwiftLogging.h [hunk 2]
```diff
@@ -23,6 +23,11 @@
 // * passing a logger around using a `Logger& logger` function parameter
 // They are created with a name that appears in the logs and can be used to filter debug levels (see
 // `Logger`).
+// If the first argument after the format is a SwiftDiagnosticSource or
```
[the remark this answers is not in the document]
Comment (redsun82) [reply swift/logging/SwiftLogging.h]: outdated comment was removed

File: swift/logging/SwiftLogging.h (hunk 1 above)
Comment (redsun82) [reply swift/logging/SwiftLogging.h:241]: using `const char*` would make it a worse overload match than the template `Args&&... args` provided by binlog, so the interception would not work. In any case, this code will go as I'm going for one of your suggestions above

**Review by sashabu — APPROVED** [review]
Thanks! :shipit:

**Pull request merged**